### ISSUE TITLE
`volatile atomics` should include `.mmio` in their lowering to uphold what CUDA C++ promises about `volatile`

### ISSUE THREAD

**Issue by jrhemstad** [NVIDIA/cccl#1613]
> `volatile atomics` should include `.mmio` in their lowering to uphold what CUDA C++ promises about `volatile`. 

 _Originally posted by @gonzalobg in [#1424](https://github.com/NVIDIA/cccl/issues/1424#issuecomment-1961770476)_

**Comment by wmaxey** [comment]
@gonzalobg @griwes 

Here's a topic up for discussion:
`volatile atomic_ref<T>` goes against the standard, however it appears that `atomic_ref<volatile T>` is allowable simply by omission. Would this be sufficient for exposing MMIO?

`volatile atomic` already serves its purpose for this as the `volatile` qualifier isn't shallow. I don't think we need to add any extra behavior there as it should end up on any volatile codepaths.

**Comment by gonzalobg** [comment]
>  atomic_ref<volatile T> is allowable simply by omission. Would this be sufficient for exposing MMIO?

Yes, but see both https://cplusplus.github.io/LWG/issue4069 and https://cplusplus.github.io/LWG/issue3508 .

**Comment by wmaxey** [comment]
> > atomic_ref is allowable simply by omission. Would this be sufficient for exposing MMIO?
> 
> Yes, but see both https://cplusplus.github.io/LWG/issue4069 and https://cplusplus.github.io/LWG/issue3508 .

Ah perfect. Thanks for the references. I would much rather implement according to 4069 as selectively stripping and adding CV qualifiers on the user API would be a nightmare.
